### PR TITLE
feat(plus): add AI Help backend

### PR DESCRIPTION
**Pull request by fiji-flo**
Adding OpenAI and Supabase integration.

### CODE REVIEW

File: .settings.test.toml [hunk 1]
```diff
@@ -45,4 +45,8 @@ basket_url = "http://localhost:4321"
 [playground]
 github_token = "foobar"
 crypt_key = "IXAe2h1MekK4LKysmMvxomja69PT6c20A3nmcDHQ2eQ="
-flag_repo = "flags"
\ No newline at end of file
+flag_repo = "flags"
+
+[ai]
+limit_reset_duration_in_sec = 5
```
Comment (caugner) [inline .settings.test.toml:51]: @fiji-flo Is this supposed to be 5 seconds for testing, or rather 300 seconds = 5 minutes?

File: src/ai/constants.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+pub const MODEL: &str = "gpt-3.5-turbo";
+pub const MODERATION_MODEL: &str = "text-embedding-ada-002";
```
Comment (LeoMcA) [inline src/ai/constants.rs:2]: Should this be `EMBEDDING_MODEL`? We don't set the moderation model, automatically taking the latest one - this constant is used in embedding.

File: src/ai/ask.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        return Err(AIError::FlaggedError);
+    }
+
+    let last_user_message = context_messages
+        .iter()
+        .last()
+        .ok_or(AIError::NoUserPrompt)?;
+
+    let related_docs =
+        get_related_docs(client, pool, last_user_message.content.replace('\n', " ")).await?;
```
Comment (LeoMcA) [inline src/ai/ask.rs:71]: What's the motivation behind removing newlines?

File: src/ai/embeddings.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+from mdn_doc_section left join mdn_doc on mdn_doc.id = mdn_doc_section.doc_id
+where length(mdn_doc_section.content) >= $4
+and (mdn_doc_section.embedding <#> $1) * -1 > $2
+order by mdn_doc_section.embedding <#> $1
+limit $3;",
+    )
+    .bind(embedding)
+    .bind(0.78)
+    .bind(3)
+    .bind(50)
```
Comment (LeoMcA) [inline src/ai/embeddings.rs:46]: perhaps these should be constants at the top of the file with descriptive names?

File: src/ai/ask.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    let mut context = vec![];
+    let mut refs = vec![];
+    let mut token_len = 0;
+    for doc in related_docs.into_iter() {
+        println!("slug: {}", doc.slug);
```
Comment (LeoMcA) [inline src/ai/ask.rs:77]: make debug

File: src/ai/helpers.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    if init_tokens + ASK_MAX_COMPLETION_TOKENS > ASK_TOKEN_LIMIT {
+        return Err(AIError::TokenLimit);
+    }
+    let mut context_tokens = num_tokens_from_messages(MODEL, &context_messages)?;
+
+    let mut index = 0;
+    while context_tokens + init_tokens + ASK_MAX_COMPLETION_TOKENS > ASK_TOKEN_LIMIT {
+        index += 1;
+        if index >= context_messages.len() {
+            return Err(AIError::TokenLimit);
```
Comment (LeoMcA) [inline src/ai/helpers.rs:41]: Should this error out, or just proceed with no historical context?

File: src/ai/ask.rs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    pub req: CreateChatCompletionRequest,
+    pub refs: Vec<RefDoc>,
+}
+
+pub async fn prepare_ask_req(
+    client: &Client<OpenAIConfig>,
+    pool: &SupaPool,
+    messages: Vec<ChatCompletionRequestMessage>,
+) -> Result<AskRequest, AIError> {
+    let open_ai_messages = sanitize_messages(messages);
```
Comment (LeoMcA) [inline src/ai/ask.rs:39]: We never use these role assistant messages, is this step necessary?

File: src/api/error.rs [hunk 8]
```diff
@@ -72,6 +75,8 @@ pub enum ApiError {
     DocumentNotFound,
     #[error("Collection with id {0} not found")]
     CollectionNotFound(String),
+    #[error("Notification Not found")]
+    NotificationNotFound,
```
Comment (LeoMcA) [inline src/api/error.rs:79]: Bad rebase?

File: src/api/error.rs [hunk 9]
```diff
@@ -114,6 +123,7 @@ impl ApiError {
             Self::DocumentNotFound => "Document not found",
             Self::InvalidBearer => "Invalid bearer info",
             Self::MalformedUrl => "Malformed URL",
+            Self::NotificationNotFound => "Notification not found",
```
Comment (LeoMcA) [inline src/api/error.rs:126]: Bad rebase?

File: tests/api/ai_help.rs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        let ask = client
+            .post(
+                "/api/v1/plus/ai/ask",
+                None,
+                Some(crate::helpers::http_client::PostPayload::Json(json!({
+                    "messages": [{ "role": "user", "content": "Foo?" }]
+                }))),
+            )
+            .await;
+        assert_eq!(ask.status(), StatusCode::NOT_IMPLEMENTED);
```
Comment (LeoMcA) [inline tests/api/ai_help.rs:42]: Would be nice to be super thorough and verify the quota counts down properly at each step in the loop.

File: tests/api/ai_help.rs [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    ))
+    .await;
+
+    let quota = client.get("/api/v1/plus/ai/ask/quota", None).await;
+    assert_ok_with_json_containing(
+        quota,
+        json!({"quota": { "count": 0, "limit": 5, "remaining": 5}}),
+    )
+    .await;
+    drop(stubr);
```
Comment (LeoMcA) [inline tests/api/ai_help.rs:135]: Would be nice to also verify we can ask again without receiving a null (or ideally rate limited) response.

File: tests/api/ai_help.rs [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+            }))),
+        )
+        .await;
+    assert_ok_with_json_containing(ask, json!(null)).await;
+    drop(stubr);
+    Ok(())
+}
+
+#[actix_rt::test]
+async fn test_quota_rest() -> Result<(), Error> {
```
Comment (LeoMcA) [inline tests/api/ai_help.rs:67]: If this does everything the above test does as a predicate, is the above test really required?

**Review by LeoMcA — COMMENTED** [review]
Nothing major, a few small things:

File: .settings.test.toml (hunk 1 above)
Comment (LeoMcA) [reply .settings.test.toml:51]: We wait this duration in the tests, so if it were 5 minutes our tests would take 5 minutes longer to run

Comment (fiji-flo) [reply .settings.test.toml:51]: Yep, that's for tests and supposed to be 5s.

File: src/db/ai.rs [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    insert_into(schema::ai_help_limits::table)
+        .values(&limit)
+        .on_conflict(schema::ai_help_limits::user_id)
+        .do_update()
+        .set(((total_questions.eq(total_questions + 1)),))
+        .execute(conn)?;
+    Ok(())
+}
+
+pub fn create_or_increment_limit(
```
[the remark this answers is not in the document]
Comment (fiji-flo) [reply src/db/ai.rs:49]: Yes, partly valid for our 24h use case. But it's okay!

File: src/ai/ask.rs (hunk 7 above)
Comment (fiji-flo) [reply src/ai/ask.rs:39]: That's to be safe. We _don't_ need it right now but we might. And this protects against injecting system messages.

File: src/ai/helpers.rs (hunk 6 above)
Comment (fiji-flo) [reply src/ai/helpers.rs:41]: Changed this from `index` to `skip`. 
This the actual last user prompt is the last message in `context_messages` if that doesn't fit we cannot do anything.

File: src/api/ai.rs [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+) -> Result<Either<impl Responder, HttpResponse>, ApiError> {
+    let mut conn = diesel_pool.get()?;
+    let user = get_user(&mut conn, user_id.id().unwrap())?;
+    let current = if user.is_subscriber() {
+        create_or_increment(&mut conn, &user)?;
+        None
+    } else {
+        let current = create_or_increment_limit(&mut conn, &user)?;
+        if current.is_none() {
+            return Ok(Either::Right(HttpResponse::Ok().json(json!(null))));
```
[the remark this answers is not in the document]
Comment (fiji-flo) [reply src/api/ai.rs:95]: Bigger decision on how we use status codes. I think right now we mostly use the payload to communicate.

File: src/db/ai.rs [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+use diesel::{insert_into, PgConnection};
+use once_cell::sync::Lazy;
+
+use crate::db::error::DbError;
+use crate::db::model::{AIHelpLimitInsert, UserQuery};
+use crate::db::schema;
+use crate::db::schema::ai_help_limits::*;
+use crate::settings::SETTINGS;
+
+pub const AI_HELP_LIMIT: i64 = 5;
```
[the remark this answers is not in the document]
Comment (fiji-flo) [reply src/db/ai.rs:12]: Happy to make it a settings once the setting isn't optional. I'd like to keep the lazy consts to minimal for now.

File: src/ai/ask.rs [hunk 16]
```diff
@@ -83,12 +84,13 @@ pub async fn prepare_ask_req(
             break;
         }
         context.push(doc.content);
-        refs.push(RefDoc {
+        ref_set.insert(RefDoc {
```
Comment (fiji-flo) [inline src/ai/ask.rs:87]: @caugner 
```rust
        if refs.iter().any(|r: &RefDoc| r.slug == doc.slug) {
            refs.push(RefDoc {
                url: doc.url,
                slug: doc.slug,
                title: doc.title,
            });
        }
```
That would have been my initial approach as it's not to many (only 3 right now).

Using a set is fine, but it this case maybe we wanna keep the order and use a [BTreeSet](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html)

**Review by LeoMcA — APPROVED** [review]
lgtm! :rocket:

File: src/ai/ask.rs (hunk 16 above)
Comment (caugner) [reply src/ai/ask.rs]: [6fdc8e5](https://github.com/mdn/rumba/pull/230/commits/6fdc8e5aae37c80dc2ff4d3d2538b7fc1dc7f861)

File: .settings.test.toml [hunk 17]
```diff
@@ -45,4 +45,8 @@ basket_url = "http://localhost:4321"
 [playground]
 github_token = "foobar"
 crypt_key = "IXAe2h1MekK4LKysmMvxomja69PT6c20A3nmcDHQ2eQ="
-flag_repo = "flags"
\ No newline at end of file
+flag_repo = "flags"
+
+[ai]
```
Comment (caugner) [inline .settings.test.toml:50]: We should also add this section to the other `.settings.*.toml` files.

Comment (fiji-flo) [reply .settings.test.toml:50]: Once they are not optional.

Comment (caugner) [reply .settings.test.toml:51]: Fixed in 54ed74c

File: migrations/2023-06-13-081720_ai-help-limits/down.sql [hunk 18]
```diff
@@ -0,0 +1 @@
+DROP TABLE ai_help_limits;
```
Comment (caugner) [inline migrations/2023-06-13-081720_ai-help-limits/down.sql:1]:
```suggestion
DROP TABLE ai_help_limits;

```

**Pull request merged**